device-server/src/main/kotlin/com/badoo/automation/deviceserver/host/AndroidEmulatorsNode.kt

```diff
@@ -0,0 +1,352 @@
+
+import com.badoo.automation.deviceserver.data.*
+import com.badoo.automation.deviceserver.host.management.errors.DeviceCreationException
+import com.badoo.automation.deviceserver.host.management.errors.DeviceNotFoundException
+import com.badoo.automation.deviceserver.host.management.errors.OverCapacityException
+import okhttp3.*
+import org.slf4j.LoggerFactory
+import java.net.HttpURLConnection
+import java.net.URI
+import java.net.URL
+import java.util.*
+import java.util.concurrent.*
+import java.util.concurrent.atomic.AtomicInteger
+import kotlin.RuntimeException
+
+typealias EmulatorSerial = String
+
+// Represents one ticket-machine = ticketHost = machine:8088.
+class AndroidEmulatorsNode(
+        private val ticketHost: String,
+        private val emulatorLimit: Int,
+        private val concurrentBoots: Int,
+        private val concurrentBooter: ExecutorService = Executors.newFixedThreadPool(concurrentBoots),
+        private val httpClient: OkHttpClient = OkHttpClient()
+) : ISimulatorsNode {
+    // Unique number for references
+    private var serial = AtomicInteger(0)
+
+    // Nicer to keep a list of known serial numbers than write a validating function, because it's similar
+    // for real devices which cannot have a validating function and may disappear if its battery dies
+    // or someone borrows one.
+    private val knownEmulators = (1..emulatorLimit).map { "emulator-${5552 + 2 * it}" }.toSet()
+
+    // Need to keep a list of allocated devices; keep ticket for handy list of existing devices.
+    private val allocatedEmulators = ConcurrentHashMap<EmulatorSerial, DeviceRef?>()
+
+    // Keep these until released by user or SessionManager
+    private val refToDto = HashMap<DeviceRef, DeviceDTO>()
+
+    private val logger = LoggerFactory.getLogger(javaClass.simpleName)
+
+    override val remoteAddress: String get() = ticketHost
+
+    override fun isReachable(): Boolean {
+        // Cheapest, quietest 200-OK no-op.
+        httpGet("docker/hello").use { return it.code() == 200 }
+    }
+
+    val newEmulatorLimit = Regex("^Now: $emulatorLimit$", RegexOption.MULTILINE)
+
+    override fun prepareNode() {
+        httpPost("docker/kill-all").use { throwUnlessOK(it) }
+        httpPost("docker/set-max-emus/$emulatorLimit")
+                .use {
+                    throwUnlessOK(it)
+                    val text = it.body()?.string()
+                    if (text == null || !text.contains(newEmulatorLimit)) {
+                        throw DeviceCreationException("Failed to set emulator max: response=$it body=$text")
+                    }
+                }
+    }
+
+    override fun deleteRelease(deviceRef: DeviceRef, reason: String): Boolean {
+        val emulator = emulatorOrNullFromRef(deviceRef) ?: return false
+        val released = releaseOnTicketMachine(deviceRef)
+        val deallocated = deallocateEmulatorSlot(emulator) != null
+        refToDto.remove(deviceRef)
+        return released && deallocated
```

Comment:
This method will return `false` in user tries to release an emulator that failed during boot and had its slot removed by `deallocateIfNotCreated`. I think it should return `true` in such a case, as release actually happened. `false` should be returned if no such emulator was found by ref or release somehow failed softly mid-ways

Reply:
I've made it return released || deallocated, and not to return early. I think that's what you meant.